File: n8n-plugin-worker/app/sandbox.mjs
/** 在一次性 Docker 容器中安装、探测或调用单个 n8n 插件。 */

import { spawn } from 'node:child_process'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { PackageError, PackageStore } from './package-store.mjs'

const root = dirname(fileURLToPath(import.meta.url))
const maximum = Math.max(1024, Math.min(Number(process.env.PLUGIN_WORKER_MAX_RESPONSE_BYTES || 16 * 1024 * 1024),
  16 * 1024 * 1024))

/** 读取 Broker 传入的有限 JSON。 */
async function readInput() {
  let value = ''; let size = 0
  for await (const chunk of process.stdin) {
    size += chunk.length
    if (size > 16 * 1024 * 1024) throw new PackageError('REQUEST_SIZE_INVALID')
    value += chunk
  }
  const parsed = JSON.parse(value)
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) throw new PackageError('REQUEST_JSON_INVALID')
  return parsed
}

/** 从当前指纹独占卷读取组件并在无长期凭据的子进程中调用。 */
async function invoke(store, request) {
  const installed = await store.metadata(String(request.fingerprint || ''))
  const component = installed.metadata.components.find(item => item.externalId === request.componentId)
  if (!component || component.compatibilityStatus === 'UNSUPPORTED') throw new PackageError('PLUGIN_COMPONENT_UNSUPPORTED')
  const payload = JSON.stringify({ ...request, root: installed.root, sourcePath: component.sourcePath,
    exportName: component.exportName, credentialAuthentications: component.credentialAuthentications || [] })
  const timeout = Math.max(1000, Math.min(Number(process.env.PLUGIN_INVOCATION_TIMEOUT_SECONDS || 60) * 1000, 300000))
  return new Promise((resolvePromise, rejectPromise) => {
    const child = spawn(process.execPath, [resolve(root, 'invoke-child.mjs')], { env: {
      PATH: process.env.PATH || '', NODE_ENV: 'production', LANG: 'C.UTF-8',
      HTTP_PROXY: process.env.HTTP_PROXY || '', HTTPS_PROXY: process.env.HTTPS_PROXY || '',
      NO_PROXY: process.env.NO_PROXY || '', PLUGIN_HTTP_RESPONSE_MAX_BYTES: process.env.PLUGIN_HTTP_RESPONSE_MAX_BYTES || '',
    }, stdio: ['pipe', 'pipe', 'pipe'] })
    let stdout = ''; let stderr = ''
    const timer = setTimeout(() => { child.kill('SIGKILL'); rejectPromise(new PackageError('PLUGIN_INVOCATION_TIMEOUT')) }, timeout)
    child.stdout.on('data', chunk => { stdout += chunk; if (stdout.length > maximum) child.kill('SIGKILL') })
    child.stderr.on('data', chunk => { stderr += chunk; if (stderr.length > 4096) child.kill('SIGKILL') })
    child.on('error', rejectPromise)
    child.on('close', code => {
      clearTimeout(timer)
      try {
        const result = JSON.parse(stdout)
        if (code !== 0 || !result.success) rejectPromise(new PackageError(result.error || 'PLUGIN_INVOCATION_FAILED'))
        else resolvePromise(result)
      } catch { rejectPromise(new PackageError('PLUGIN_OUTPUT_INVALID')) }
    })
    child.stdin.end(payload)
  })
}

try {
  const operation = process.argv[2]
  if (!['inspect', 'invoke'].includes(operation)) throw new PackageError('PLUGIN_SANDBOX_OPERATION_INVALID')
  const request = await readInput()
  const store = new PackageStore()
  const result = operation === 'inspect' ? await store.install(request) : await invoke(store, request)
  process.stdout.write(JSON.stringify(result))
} catch (error) {
  const code = error instanceof PackageError && /^[A-Z0-9_]{1,80}$/.test(error.message)
    ? error.message : 'PLUGIN_SANDBOX_FAILURE'
  process.stdout.write(JSON.stringify({ error: code }))
  process.exitCode = 1
}

File: dify-plugin-worker/app/sandbox.py
"""在一次性 Docker 容器中安装、探测或调用单个 Dify 插件。"""

from __future__ import annotations

import json
import os
import subprocess
import sys
from typing import Any

from app.package_store import PackageError, PackageStore


def invoke(store: PackageStore, request: dict[str, Any]) -> dict[str, Any]:
    """从当前指纹独占卷读取组件并在无长期凭据的子进程中调用。"""
    root, metadata = store.metadata(str(request.get("fingerprint", "")))
    external_id = str(request.get("componentId", ""))
    component = next((item for item in metadata["components"] if item["externalId"] == external_id), None)
    if component is None or component["compatibilityStatus"] == "UNSUPPORTED":
        raise ValueError("PLUGIN_COMPONENT_UNSUPPORTED")
    child_request = dict(request)
    child_request["root"] = str(root)
    child_request["sourcePath"] = component["sourcePath"]
    child_request["componentType"] = component.get("componentType", "")
    child_request["modelType"] = component.get("modelType", "")
    environment = {
        "PATH": os.getenv("PATH", ""), "PYTHONPATH": "/app", "PYTHONDONTWRITEBYTECODE": "1",
        "PYTHONUNBUFFERED": "1", "LANG": "C.UTF-8", "HTTP_PROXY": os.getenv("HTTP_PROXY", ""),
        "HTTPS_PROXY": os.getenv("HTTPS_PROXY", ""), "NO_PROXY": os.getenv("NO_PROXY", ""),
    }
    timeout = max(1, min(int(os.getenv("PLUGIN_INVOCATION_TIMEOUT_SECONDS", "60")), 300))
    try:
        result = subprocess.run([sys.executable, "-m", "app.invoke_child"], input=json.dumps(child_request),
                                text=True, capture_output=True, timeout=timeout, env=environment, check=False)
    except subprocess.TimeoutExpired as exception:
        raise TimeoutError("PLUGIN_INVOCATION_TIMEOUT") from exception
    try:
        response = json.loads(result.stdout)
    except Exception as exception:
        raise ValueError("PLUGIN_OUTPUT_INVALID") from exception
    if result.returncode != 0 or not response.get("success"):
        raise ValueError(str(response.get("error") or "PLUGIN_INVOCATION_FAILED"))
    return response


def main() -> None:
    """读取 Broker 标准输入并执行命令行固定的单一沙箱操作。"""
    if len(sys.argv) != 2 or sys.argv[1] not in {"inspect", "invoke"}:
        raise ValueError("PLUGIN_SANDBOX_OPERATION_INVALID")
    request = json.loads(sys.stdin.read())
    if not isinstance(request, dict):
        raise ValueError("REQUEST_JSON_INVALID")
    store = PackageStore()
    result = store.install(request) if sys.argv[1] == "inspect" else invoke(store, request)
    sys.stdout.write(json.dumps(result, ensure_ascii=False, separators=(",", ":")))


if __name__ == "__main__":
    try:
        main()
    except (PackageError, TimeoutError, ValueError) as exception:
        sys.stdout.write(json.dumps({"error": str(exception)[:500]}, ensure_ascii=False))
        sys.exit(1)
    except Exception:
        sys.stdout.write(json.dumps({"error": "PLUGIN_SANDBOX_FAILURE"}))
        sys.exit(1)
